plugins/module_utils/foreman_helper.py

```diff
@@ -590,10 +591,238 @@ def exit_json(self, changed=False, **kwargs):
         super(ForemanAnsibleModule, self).exit_json(**kwargs)
 
 
-class ForemanEntityAnsibleModule(ForemanAnsibleModule):
+class ResolverEntityMixin(object):
+    """ This mixin add automatic entities resolution based on provided attributes/ sub entities options.
+        It add following options to entity_spec 'entity' and 'entity_list' types:
+        * search_by (str): Field used to search the sub entity. Defaults to 'name'.
+        * search_operator (str): Operator used to search the su entity. Defaults to '='.
+        * resource_type (str): Resource type used to build API resource PATH. Defaults to pluralized entity key.
+        It add following attributes:
+        * entity_search (str): Defautls to None. If provided, the base entity resolver will use this search query instead to try to build it.
+        * entity_key (str): field used to search current entity. Defaults to 'name'. If entityèspec contains 'parent', defaults to 'title'.
+        * entity_scope (str): Type of the entity used to build search scope. Defaults to None.
+        * entity_resolve (boolean): Set it to False to disable base entity resolution.
+        * entity_opts (dict): Dict of options for base entity. Same options can be provided for subentities described in entity_spec.
+    """
+    def __init__(self, argument_spec=None, **kwargs):
+        self.original_entity_spec = kwargs.pop('entity_spec', {})
+        self.sub_entities = {key: value for key, value in self.original_entity_spec.items()
+                             if value.get('resolve', True) and 'type' in value and value['type'] in ['entity', 'entity_list']}
+        self.entity_scope = kwargs.pop('entity_scope', None)
+        self.entity_key = kwargs.pop('entity_key', 'name')
+        self.entity_resolve = kwargs.pop('entity_resolve', True)
+        self.entity_opts = kwargs.pop('entity_opts', {})
+        if 'thin' not in self.entity_opts:
+            self.entity_opts['thin'] = False
+        if 'failsafe' not in self.entity_opts:
+            self.entity_opts['failsafe'] = True
+        if 'parent' in self.sub_entities:
+            self.entity_opts['search_by'] = 'title'
+        elif 'search_by' not in self.entity_opts:
+            self.entity_opts['search_by'] = self.entity_key
+
+        # Convert current class name from CamelCase to snake_case
+        class_name = re.sub(r'(?<=[a-z])[A-Z]|[A-Z](?=[^A-Z])', r'_\g<0>', self.__class__.__name__).lower().strip('_')
+        # Get entity name from snake case class name
+        self.entity_name = '_'.join(class_name.split('_')[1:-1])
+        p = apypie.Inflector()
+        self._entity_resource_path = p.pluralize(self.entity_name)
+        self.search = ""
+        super(ResolverEntityMixin, self).__init__(argument_spec=argument_spec, **kwargs)
+
+    def run(self, entity_dict=None, entity=None, connect=True, exit=True):
```

Comment:
I think I once talked with either @mdellweg or @ekohl about this (the semi-automatic modules), and we came up with the idea to have `connect()` as a context manager, so the api would be smth like:

```python
with module.connect():
    module.run()
```

this would allow us to avoid the `connect` and `exit` params and easily add steps before and after `run`.

What do you think?

Reply:
If you meant the "there is no propoper destructor idiom in python" discussion, it was me. And i'd vote for it.

Reply:
I don't recall talking about this, but it looks very nice.

Reply:
Yeah, it was Matthias :)

Reply:
Maybe make this in another PR ? This one is already a big one... I submit a PR with the context manager and rebase this one on it when merged ? What do you think ?

Reply:
@evgeni, @mdellweg, @ekohl I've added a context manager called `start()` (the name can change if you want, lack of inspiration)